packages/backend/src/models/user.ts

```diff
@@ -148,40 +148,136 @@ class User extends Base {
+            .where('flow_collaborators.user_id', userId)
+            .whereIn('flow_collaborators.role', ['editor', 'viewer'])
+            .whereNull('flow_collaborators.deleted_at')
+        })
+      })
+    }
+  }
+
+  withAccessible(params: {
+    type: 'connection'
+    queryBuilder?: ExtendedQueryBuilder<Connection, Connection[]>
+    trx?: Transaction
+    requiredRole: IFlowCollabRole
+  }): ExtendedQueryBuilder<Connection, Connection[]>
+  withAccessible(params: {
+    type: 'flow'
+    queryBuilder?: ExtendedQueryBuilder<Flow, Flow[]>
+    trx?: Transaction
+    requiredRole: IFlowCollabRole
+  }): ExtendedQueryBuilder<Flow, Flow[]>
+  withAccessible(params: {
+    type: 'step'
+    queryBuilder?: ExtendedQueryBuilder<Step, Step[]>
+    trx?: Transaction
+    requiredRole: IFlowCollabRole
+  }): ExtendedQueryBuilder<Step, Step[]>
+  withAccessible({
+    type,
     queryBuilder,
     trx,
+    requiredRole,
   }: {
-    queryBuilder?: ExtendedQueryBuilder<Flow, Flow[]>
+    type: 'flow' | 'step' | 'connection'
+    queryBuilder?:
+      | ExtendedQueryBuilder<Flow, Flow[]>
+      | ExtendedQueryBuilder<Step, Step[]>
+      | ExtendedQueryBuilder<Connection, Connection[]>
     trx?: Transaction
-  } = {}) {
+    requiredRole: IFlowCollabRole
+  }) {
     const userId = this.id
-    const baseQuery = queryBuilder || Flow.query(trx)
-    return baseQuery
-      .select(
-        'flows.*',
-        Flow.raw(
-          `CASE
+    const USER_ROLE_STMT = `
+      CASE
+
+
+
+
         WHEN flows.user_id = ? THEN 'owner'
         ELSE (
           SELECT role FROM flow_collaborators
           WHERE flow_collaborators.flow_id = flows.id
           AND flow_collaborators.user_id = ?
           AND flow_collaborators.deleted_at IS NULL
         )
-      END as role`,
-          [userId, userId],
-        ),
-      )
-      .where(function () {
-        this.where('flows.user_id', userId).orWhereExists(function () {
-          this.select(1)
-            .from('flow_collaborators')
-            .whereRaw('flow_collaborators.flow_id = flows.id')
-            .where('flow_collaborators.user_id', userId)
-            .whereNull('flow_collaborators.deleted_at')
-        })
-      })
+      END as role
+    `
+
+    let baseQuery:
+      | ExtendedQueryBuilder<Flow, Flow[]>
+      | ExtendedQueryBuilder<Step, Step[]>
+      | ExtendedQueryBuilder<Connection, Connection[]>
+
+    switch (type) {
+      case 'connection':
+        baseQuery =
+          (queryBuilder as ExtendedQueryBuilder<Connection, Connection[]>) ||
+          Connection.query(trx)
+        baseQuery
+          .select(
+            'connections.*',
+            Connection.raw(USER_ROLE_STMT, [userId, userId]),
+          )
+          .join('steps', 'steps.connection_id', 'connections.id')
+          .join('flows', 'steps.flow_id', 'flows.id')
+          .leftJoin('flow_connections', 'flow_connections.flow_id', 'flows.id')
```

Comment:
is there a need to join? can be it fetched from user's connections as well as flow_connections